src/main/java/net/minecraftforge/common/smelting/SmeltingHandler.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2018.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.smelting;
+
+import com.google.common.collect.Maps;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.FurnaceRecipes;
+import net.minecraftforge.fml.common.FMLLog;
+import net.minecraftforge.oredict.OreDictionary;
+
+import java.util.Map;
+
+public class SmeltingHandler extends FurnaceRecipes
+{
+    private static final SmeltingHandler INSTANCE = new SmeltingHandler();
+
+    private Map<ItemStack, ISmeltingRecipe> recipes;
+    private Map<ItemStack, ItemStack> staticView;
+
+    private SmeltingHandler()
+    {
+        super();
+    }
+
+    public static SmeltingHandler instance()
+    {
+        return SmeltingHandler.INSTANCE;
+    }
+
+    @Override
+    public void addSmeltingRecipe(ItemStack input, ItemStack stack, float experience)
+    {
+        addSmeltingRecipe(input, new BasicSmeltingRecipe(stack, experience));
+    }
+
+    public void addSmeltingRecipe(ItemStack input, ISmeltingRecipe recipe)
+    {
+        if (findRecipe(input) != null)
+        {
+            FMLLog.log.info("Ignored smelting recipe with conflicting input: {} = {}", input, recipe.getGenericOutput());
+            return;
+        }
+        getRecipes().put(input, recipe);
+    }
+
+    @Override
+    public ItemStack getSmeltingResult(ItemStack stack)
+    {
+        ISmeltingRecipe recipe = findRecipe(stack);
+        if (recipe != null)
+            return recipe.getSmeltingResult(stack);
+        else
+            return ItemStack.EMPTY;
+    }
+
+    @Override
+    public Map<ItemStack, ItemStack> getSmeltingList()
+    {
+        return staticView;
+    }
+
+    public Map<ItemStack, ISmeltingRecipe> getRecipes()
+    {
+        if (recipes != null)
+            return recipes;
+        recipes = Maps.newHashMap();
```

Comment:
I suggest using array-backed maps instead because the map is always iterated over (itemstack class does not override hashcode or equals)

Reply:
using an array-backed map would probably make sense - as for being able to intialize it in the constructor: it isnt possible as the super-constructor calls overridden methods that require the field set